File: apps/website/frontend/src/features/tactical-map/worldmap/chunkMath.ts
// ORACLE-ONLY (T-151.11.3, audit B-03): the live paths call wasm `chunk_ids_for_viewport`
// / `WorldResidency.set_viewport`; this file feeds parity/unit tests only.
// T-090.5.1 — Export-chunk math for Map Engine v2 streaming (implementation plan §6). Pure:
// no React/Deck/fetch, node-testable; the worker (T-090.5.3) and chunkStore consume it.
//
// Chunk key contract (T-090.3.1 export): (cx, cy) = floor(x/512), floor(y/512) in world meters,
// file names `objects/chunks/{cx}_{cy}.json.gz` + `objects/density/{cx}_{cy}.bin`. The runtime
// authority for the cell size is manifest `objects.chunkSizeM`; every function takes an optional
// override and defaults to 512. NOTE: state/spatialChunks.ts also bins at 512 m — that grid culls
// mission SLOT icons (T-067) and is a separate domain; do not merge the two.

/** Default export chunk edge in world meters (= manifest `objects.chunkSizeM` today). */
export const DEFAULT_CHUNK_SIZE_M = 512

/** World bbox [minX, minY, maxX, maxY] in meters. */
export type Bbox = [number, number, number, number]

/** Inclusive chunk-index rectangle. */
export interface ChunkRect {
  cx0: number
  cy0: number
  cx1: number
  cy1: number
}

/** Terrain extent in world meters (Everon 12800×12800). */
export interface TerrainSizeM {
  width: number
  height: number
}

const clampInt = (v: number, lo: number, hi: number) => Math.max(lo, Math.min(hi, v))

/** Export-artifact id for a chunk — matches the on-disk `{cx}_{cy}` file stem. */
export function chunkId(cx: number, cy: number): string {
  return `${cx}_${cy}`
}

/** Border-preload margin in meters (plan §6, A3 PreloadMapObjects analogue):
 *  max(5% of the larger viewport span, one chunk ring). */
export function preloadMarginM(bbox: Bbox, chunkSizeM = DEFAULT_CHUNK_SIZE_M): number {
  const span = Math.max(bbox[2] - bbox[0], bbox[3] - bbox[1])
  return Math.max(0.05 * span, chunkSizeM)
}

/** Expand a bbox by a symmetric margin (no clamping — chunk conversion clamps). */
export function expandBbox(bbox: Bbox, marginM: number): Bbox {
  return [bbox[0] - marginM, bbox[1] - marginM, bbox[2] + marginM, bbox[3] + marginM]
}

/** Bbox → inclusive chunk rect, clamped to the terrain grid so we never request ids that
 *  cannot exist on disk. Degenerate/edge coordinates clamp to the last row/column. */
export function chunkRectForBbox(
  bbox: Bbox,
  terrain: TerrainSizeM,
  chunkSizeM = DEFAULT_CHUNK_SIZE_M,
): ChunkRect {
  const maxCx = Math.max(0, Math.ceil(terrain.width / chunkSizeM) - 1)
  const maxCy = Math.max(0, Math.ceil(terrain.height / chunkSizeM) - 1)
  return {
    cx0: clampInt(Math.floor(Math.min(bbox[0], bbox[2]) / chunkSizeM), 0, maxCx),
    cy0: clampInt(Math.floor(Math.min(bbox[1], bbox[3]) / chunkSizeM), 0, maxCy),
    cx1: clampInt(Math.floor(Math.max(bbox[0], bbox[2]) / chunkSizeM), 0, maxCx),
    cy1: clampInt(Math.floor(Math.max(bbox[1], bbox[3]) / chunkSizeM), 0, maxCy),
  }
}

/** Grow a chunk rect by `ring` chunks on every side, clamped to the terrain grid. Used for the
 *  oversized-object +1 ring (plan §6: runway/pier/powerline spans exceed their home chunk). */
export function expandChunkRect(
  rect: ChunkRect,
  ring: number,
  terrain: TerrainSizeM,
  chunkSizeM = DEFAULT_CHUNK_SIZE_M,
): ChunkRect {
  const maxCx = Math.max(0, Math.ceil(terrain.width / chunkSizeM) - 1)
  const maxCy = Math.max(0, Math.ceil(terrain.height / chunkSizeM) - 1)
  return {
    cx0: clampInt(rect.cx0 - ring, 0, maxCx),
    cy0: clampInt(rect.cy0 - ring, 0, maxCy),
    cx1: clampInt(rect.cx1 + ring, 0, maxCx),
    cy1: clampInt(rect.cy1 + ring, 0, maxCy),
  }
}

/** Enumerate a rect's chunk ids row-major (stable order → stable fetch/dedupe keys). */
export function chunkIdsForRect(rect: ChunkRect): string[] {
  const ids: string[] = []
  for (let cy = rect.cy0; cy <= rect.cy1; cy++) {
    for (let cx = rect.cx0; cx <= rect.cx1; cx++) ids.push(chunkId(cx, cy))
  }
  return ids
}

/** Viewport bbox → chunk ids to hydrate: bbox grown by the preload margin (§6 border rule),
 *  optionally + `extraRing` chunks (oversized-object classes), clamped to the terrain. */
export function chunkIdsForViewport(
  bbox: Bbox,
  terrain: TerrainSizeM,
  opts?: { chunkSizeM?: number; extraRing?: number },
): string[] {
  const chunkSizeM = opts?.chunkSizeM ?? DEFAULT_CHUNK_SIZE_M
  const preloaded = expandBbox(bbox, preloadMarginM(bbox, chunkSizeM))
  let rect = chunkRectForBbox(preloaded, terrain, chunkSizeM)
  if (opts?.extraRing) rect = expandChunkRect(rect, opts.extraRing, terrain, chunkSizeM)
  return chunkIdsForRect(rect)
}

File: apps/website/graphics-engine/src/streaming/scheduler/chunk_math.rs
//! Role: chunk math.
//! Position: `streaming/scheduler` in the graphics engine.
//! Signals & state: camera, spatial, asset, or GPU data owned by this module.
//! Invariants: preserve coordinates, resource lifetimes, ordering, and binary layouts.

/// A world bbox `[minX, minY, maxX, maxY]` in meters (the `chunkMath.ts` `Bbox`).
pub type Bbox = [f64; 4];

/// Terrain extent in meters (`chunkMath.ts` `TerrainSizeM`). `Default` = `0×0` (an unconfigured residency then resolves every viewport to the empty set until a manifest sets the bounds).
#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct TerrainSizeM {
    /// Width.
    pub width: f64,

    /// Height.
    pub height: f64,
}

/// Inclusive chunk-index rect (`chunkMath.ts` `ChunkRect`).
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct ChunkRect {
    /// Cx0.
    pub cx0: i64,

    /// Cy0.
    pub cy0: i64,

    /// Cx1.
    pub cx1: i64,

    /// Cy1.
    pub cy1: i64,
}

/// `chunkId(cx, cy)` — `` `${cx}_${cy}` ``.
#[must_use]
pub fn chunk_id(cx: i64, cy: i64) -> String {
    format!("{cx}_{cy}")
}

#[inline]
fn clamp_f(v: f64, lo: f64, hi: f64) -> f64 {
    lo.max(hi.min(v))
}

/// `preloadMarginM(bbox, chunkSizeM)` — `max(0.05 * span, chunkSizeM)`, `span = max(w, h)`.
#[must_use]
pub fn preload_margin_m(bbox: Bbox, chunk_size_m: f64) -> f64 {
    let span = (bbox[2] - bbox[0]).max(bbox[3] - bbox[1]);
    (0.05 * span).max(chunk_size_m)
}

/// `expandBbox(bbox, marginM)` — grow every edge outward by `margin_m` (no clamp).
#[must_use]
pub fn expand_bbox(bbox: Bbox, margin_m: f64) -> Bbox {
    [
        bbox[0] - margin_m,
        bbox[1] - margin_m,
        bbox[2] + margin_m,
        bbox[3] + margin_m,
    ]
}

/// `chunkRectForBbox(bbox, terrain, chunkSizeM)`. `maxC = max(0, ceil(dim/size) - 1)`; each edge = `clampInt(floor(coord/size), 0, maxC)` with `coord = min|max(bbox edges)`.
#[must_use]
pub fn chunk_rect_for_bbox(bbox: Bbox, terrain: TerrainSizeM, chunk_size_m: f64) -> ChunkRect {
    let max_cx = 0.0_f64.max((terrain.width / chunk_size_m).ceil() - 1.0);
    let max_cy = 0.0_f64.max((terrain.height / chunk_size_m).ceil() - 1.0);
    let cx0 = clamp_f((bbox[0].min(bbox[2]) / chunk_size_m).floor(), 0.0, max_cx);
    let cy0 = clamp_f((bbox[1].min(bbox[3]) / chunk_size_m).floor(), 0.0, max_cy);
    let cx1 = clamp_f((bbox[0].max(bbox[2]) / chunk_size_m).floor(), 0.0, max_cx);
    let cy1 = clamp_f((bbox[1].max(bbox[3]) / chunk_size_m).floor(), 0.0, max_cy);
    ChunkRect {
        cx0: cx0 as i64,
        cy0: cy0 as i64,
        cx1: cx1 as i64,
        cy1: cy1 as i64,
    }
}

/// `expandChunkRect(rect, ring, terrain, chunkSizeM)` — grow the rect by `ring` chunks per edge, each edge `clampInt(_, 0, maxC)`.
#[must_use]
pub fn expand_chunk_rect(
    rect: ChunkRect,
    ring: i64,
    terrain: TerrainSizeM,
    chunk_size_m: f64,
) -> ChunkRect {
    let max_cx = 0.0_f64.max((terrain.width / chunk_size_m).ceil() - 1.0) as i64;
    let max_cy = 0.0_f64.max((terrain.height / chunk_size_m).ceil() - 1.0) as i64;
    ChunkRect {
        cx0: (rect.cx0 - ring).clamp(0, max_cx),
        cy0: (rect.cy0 - ring).clamp(0, max_cy),
        cx1: (rect.cx1 + ring).clamp(0, max_cx),
        cy1: (rect.cy1 + ring).clamp(0, max_cy),
    }
}

/// `chunkIdsForRect(rect)` — **row-major, cy outer, cx inner** (the fetch/dedupe/pin order).
#[must_use]
pub fn chunk_ids_for_rect(rect: ChunkRect) -> Vec<String> {
    let mut out = Vec::new();
    for cy in rect.cy0..=rect.cy1 {
        for cx in rect.cx0..=rect.cx1 {
            out.push(chunk_id(cx, cy));
        }
    }
    out
}

/// `chunkIdsForViewport(bbox, terrain, { chunkSizeM, extraRing })`. `extra_ring = 0` skips the oversized expansion (the JS `opts.extraRing` default).
#[must_use]
pub fn chunk_ids_for_viewport(
    bbox: Bbox,
    terrain: TerrainSizeM,
    chunk_size_m: f64,
    extra_ring: i64,
) -> Vec<String> {
    let preloaded = expand_bbox(bbox, preload_margin_m(bbox, chunk_size_m));
    let mut rect = chunk_rect_for_bbox(preloaded, terrain, chunk_size_m);
    if extra_ring > 0 {
        rect = expand_chunk_rect(rect, extra_ring, terrain, chunk_size_m);
    }
    chunk_ids_for_rect(rect)
}

#[cfg(test)]
#[path = "tests/chunk_math_tests.rs"]
mod tests;
